cpp/ql/test/experimental/query-tests/Security/CWE/CWE-415/semmle/tests/test.c

```diff
@@ -1,21 +1,96 @@
 typedef unsigned long size_t;
 void *malloc(size_t size);
 void free(void *ptr);
+#define NULL 0
 
 void workFunction_0(char *s) {
   int intSize = 10;
   char *buf;
   buf = (char *) malloc(intSize);
-  free(buf); // BAD
-  if(buf) free(buf);
+  free(buf); // GOOD
+  if(buf) free(buf); // BAD
 }
 void workFunction_1(char *s) {
+  int intSize = 10;
+  char *buf;
+  buf = (char *) malloc(intSize);
+  free(buf); // GOOD
+  free(buf); // BAD
+}
+void workFunction_2(char *s) {
+  int intSize = 10;
+  char *buf;
+  buf = (char *) malloc(intSize);
+  free(buf); // GOOD
+  buf = NULL;
+  free(buf);
+}
+void workFunction_3(char *s) {
+  int intSize = 10;
+  char *buf;
+  int intFlag;
+  buf = (char *) malloc(intSize);
+  if(buf[1]%5) {
+    free(buf);
+    buf = NULL;
+  }
+  free(buf);
+}
+void workFunction_4(char *s) {
+  int intSize = 10;
+  char *buf;
+  char *tmpbuf;
+  tmpbuf = (char *) malloc(intSize);
+  buf = (char *) malloc(intSize);
+  free(buf); // GOOD
+  buf = tmpbuf;
+  free(buf); // GOOD
+}
+void workFunction_5(char *s) {
+  int intSize = 10;
+  char *buf;
+  int intFlag;
+  buf = (char *) malloc(intSize);
+  if(intFlag) {
+    free(buf);
+  }
+  free(buf); // BAD
```

Comment:
How it is different from `workFunction_3` case?

Reply:
Good day.
the difference is that assigning a null value to a pointer eliminates the double-free error.

Reply:
Missed that. Good!